src/app/utilities/moodle/helper.ts

```diff
@@ -0,0 +1,175 @@
+import {
+  MoodleCourse,
+  MoodleException,
+  MoodleUser,
+  MoodleUserBasic,
+} from '@/app/interfaces/Moodle';
+import { getSearchQuery, getUniqueID } from '../common';
+import {
+  MOODLE_API_URL,
+  MOODLE_HOST_URL,
+  MOODLE_SECRET,
+  MOODLE_STUDENT_ROLE_ID,
+} from './constants';
+import isAuthenticated from '../auth/isAuthenticated';
+import { NextRequest } from 'next/server';
+import AuthErrorResponse from '@/app/interfaces/AuthErrorResponse';
+
+export async function getMoodleUserByEmail(
+  email: string
+): Promise<MoodleUserBasic | null> {
+  try {
+    const url = `${MOODLE_API_URL}?
+        ${getSearchQuery({
+          wstoken: MOODLE_SECRET,
+          wsfunction: 'core_user_get_users_by_field',
+          field: 'email',
+          'values[]': email,
+          moodlewsrestformat: 'json',
+        })}`;
+
+    const res = await fetch(url);
+    const json: MoodleUserBasic[] | MoodleException = await res.json();
+    if ('exception' in json) {
+      throw new Error(json.message);
+    }
+
+    const user: MoodleUserBasic = json[0];
+
+    return user;
+  } catch (_) {
+    throw new Error('Failed to fetch the moodle user.');
+  }
+}
+
+export async function createMoodleUser({
+  email,
+  name,
+}: {
+  email: string;
+  name: string;
+}): Promise<MoodleUser> {
+  try {
+    const splitName = name.split(/\s+/);
+    const firstname = splitName.slice(0, -1).join(' ');
+    const lastname = splitName.slice(-1).join('');
+    const username = email;
+
+    const password = 'TestPass123!' + getUniqueID();
+
+    const formData = new FormData();
+    formData.append('wstoken', MOODLE_SECRET);
+    formData.append('wsfunction', 'core_user_create_users');
+    formData.append('moodlewsrestformat', 'json');
+    formData.append('users[0][email]', email);
+    formData.append('users[0][firstname]', firstname);
+    formData.append('users[0][lastname]', lastname);
+    formData.append('users[0][password]', password);
+    formData.append('users[0][username]', username);
+    formData.append('users[0][auth]', 'oidc');
```

Comment:
There must be a way to create a type for the formData here and attach this object. So that we can enforce some sort of rules.


Something to do in the future.
Is there a way to decouple this? seems like it can accept bulk user creations. Can we pass it an object for it to determine whether it is single or more users?